engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java

```diff
@@ -60,26 +66,33 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
       logger.trace("Class {} found in cache", name);
       return l;
     }
-    if (!isRuntimeModInBootLayer && name.startsWith("org.graalvm")) {
-      return polyglotClassLoader.loadClass(name);
-    }
-    if (name.startsWith("org.slf4j")) {
-      // Delegating to system class loader ensures that log classes are not loaded again
-      // and do not require special setup. In other words, it is using log configuration that
-      // has been setup by the runner that started the process. See #11641.
-      return polyglotClassLoader.loadClass(name);
-    }
-    try {
-      l = findClass(name);
-      if (resolve) {
-        l.getMethods();
+    synchronized (INSTANCE) {
+      l = loadedClasses.get(name);
+      if (l != null) {
+        logger.trace("Class {} found in cache", name);
+        return l;
+      }
+      if (!isRuntimeModInBootLayer && name.startsWith("org.graalvm")) {
+        return polyglotClassLoader.loadClass(name);
+      }
+      if (name.startsWith("org.slf4j")) {
+        // Delegating to system class loader ensures that log classes are not loaded again
+        // and do not require special setup. In other words, it is using log configuration that
+        // has been setup by the runner that started the process. See #11641.
+        return polyglotClassLoader.loadClass(name);
+      }
```

Comment:
- What can be the source of concurrency problems? I am asking myself
- probably just the read and write to `loadedClasses`, right?
- This can be **moved out** of the `synchronized` block then

Reply:
- OK, thanks for sharing the failure: `attempted duplicate class definition for org.enso.base.Text_Utils`
- As far I can interpret it, it is related to two threads calling
   - into `URLClassLoader.findClass` for the same `name`
   - which then calls `defineClass` twice
   - and that throws the exception on the second attempt
- good to know the `synchronized` block fixes the problem
- it is just _too wide_ synchronization ...
- ...using the [one with Futures](https://github.com/enso-org/enso/pull/13219#discussion_r2131542150) would be more fine grain
- but such a change can certainly wait for now - it'd be good to have it before integration however